src/astarte_message_hub.rs

```diff
@@ -71,6 +85,182 @@ where
             astarte_handler,
         }
     }
+
+    /// Create a [tower] layer to intercept the Node ID in the gRPC requests
+    pub fn make_interceptor_layer(&self) -> NodeIdInterceptorLayer {
+        NodeIdInterceptorLayer::new(Arc::clone(&self.nodes))
+    }
+}
+
+/// Errors related to the interception of the NodeId into gRPC request
+#[derive(thiserror::Error, Debug, displaydoc::Display)]
+pub enum InterceptorError {
+    /// Node ID not specified in the gRPC request
+    AbsentId,
+    /// Invalid Uuid bytes value for Node ID, {0}
+    Uuid(#[from] uuid::Error),
+    /// Invalid Uuid str value for Node ID, {0}
+    ParseStr(#[from] tonic::metadata::errors::ToStrError),
+    /// The Node ID does not exist
+    IdNotFound,
+    /// Couldn't decode metadata binary value
+    MetadataBytes(#[from] InvalidMetadataValueBytes),
+    /// Error returned by the inner service
+    #[error(transparent)]
+    Inner(#[from] Box<dyn std::error::Error + Send + Sync + 'static>),
+}
+
+#[derive(Clone, Default)]
+pub struct NodeIdInterceptorLayer {
+    msg_hub_nodes: Arc<RwLock<HashMap<Uuid, AstarteNode>>>,
+}
+
+impl NodeIdInterceptorLayer {
+    fn new(msg_hub_nodes: Arc<RwLock<HashMap<Uuid, AstarteNode>>>) -> NodeIdInterceptorLayer {
+        Self { msg_hub_nodes }
+    }
+}
+
+impl<S> Layer<S> for NodeIdInterceptorLayer {
+    type Service = NodeIdInterceptor<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        NodeIdInterceptor {
+            inner: service,
+            msg_hub_nodes: self.msg_hub_nodes.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct NodeIdInterceptor<S> {
+    inner: S,
+    msg_hub_nodes: Arc<RwLock<HashMap<Uuid, AstarteNode>>>,
+}
+
+impl<S> NodeIdInterceptor<S> {
+    pub async fn handle_req(
+        mut self,
+        req: hyper::Request<Body>,
+    ) -> Result<S::Response, InterceptorError>
+    where
+        S: Service<hyper::Request<Body>, Response = hyper::Response<BoxBody>>
+            + Clone
+            + Send
+            + 'static,
+        S::Future: Send + 'static,
+        S::Error: Into<InterceptorError>,
+    {
+        static ATTACH_PATH: Lazy<http::uri::PathAndQuery> = Lazy::new(|| {
+            http::uri::PathAndQuery::from_static("/astarteplatform.msghub.MessageHub/Attach")
+        });
+
+        // check that the Node ID is present inside the metadata on all gRPC requests apart from the
+        // attach ones
+        if req.uri().path_and_query() == Some(&ATTACH_PATH) {
```

Comment:
We should just check the path and not the query parameters here

```suggestion
        if req.uri().path() == "/astarteplatform.msghub.MessageHub/Attach" {
```

Reply:
I probably would extract this check in a function `is_attach`, so we can check that the path is resolved correctly in the tests.

For example 

```rust 
let uri = "/astarteplatform.msghub.MessageHub//Attach?foo=bar".parse::<Uri>().unwrap();
assert!(is_attach(uri))
```